api/v1alpha1/physical_machine_chaos_webhook.go

```diff
@@ -77,5 +81,317 @@ func (in *PhysicalMachineChaosSpec) Validate(root interface{}, path *field.Path)
+	if len(spec.Device) == 0 {
+		return errors.New("device is required")
+	}
+
+	if len(spec.IPAddress) == 0 && len(spec.Hostname) == 0 {
+		return errors.New("one of ip-address and hostname is required")
+	}
+
+	if spec.Direction != "to" && spec.Direction != "from" {
+		return errors.New("direction should be one of 'to' and 'from'")
+	}
+
+	if len(spec.AcceptTCPFlags) > 0 && spec.IPProtocol != "tcp" {
+		return errors.Errorf("protocol should be 'tcp' when set accept-tcp-flags")
+	}
+
+	return nil
+}
+
+func validateNetworkDNSAction(spec *NetworkDNSSpec) error {
+	if (len(spec.DNSDomainName) != 0 && len(spec.DNSIp) == 0) || (len(spec.DNSDomainName) == 0 && len(spec.DNSIp) != 0) {
+		return errors.Errorf("DNS host %s must match a DNS ip %s", spec.DNSDomainName, spec.DNSIp)
+	}
+
+	return nil
+}
+
+func validateProcessAction(spec *ProcessSpec) error {
+	if len(spec.Process) == 0 {
+		return errors.New("process is required")
+	}
+
+	if spec.Signal == 0 {
+		return errors.New("signal is required")
+	}
+
+	return nil
+}
+
+func validateJVMClassMethod(spec *JVMClassMethodSpec) error {
+	if len(spec.Class) == 0 {
+		return errors.New("class is required")
+	}
+
+	if len(spec.Method) == 0 {
+		return errors.New("method is required")
+	}
+
+	return nil
+}
+
+func validateJVMExceptionAction(spec *JVMExceptionSpec) error {
+	if spec.Pid == 0 {
+		return errors.New("pid is required")
+	}
+
+	if err := validateJVMClassMethod(&spec.JVMClassMethodSpec); err != nil {
+		return err
+	}
+
+	if len(spec.ThrowException) == 0 {
+		return errors.New("exception is required")
+	}
+
+	return nil
+}
+
+func validateJVMGCAction(spec *JVMGCSpec) error {
+	if spec.Pid == 0 {
+		return errors.New("pid is required")
+	}
+
+	return nil
+}
+
+func validateJVMLatencyAction(spec *JVMLatencySpec) error {
+	if spec.Pid == 0 {
+		return errors.New("pid is required")
+	}
+
+	if err := validateJVMClassMethod(&spec.JVMClassMethodSpec); err != nil {
+		return err
+	}
+
+	if spec.LatencyDuration == 0 {
+		return errors.New("latency is required")
+	}
+
+	return nil
+}
+
+func validateJVMReturnAction(spec *JVMReturnSpec) error {
+	if spec.Pid == 0 {
+		return errors.New("pid is required")
+	}
+
+	if err := validateJVMClassMethod(&spec.JVMClassMethodSpec); err != nil {
+		return err
+	}
+
+	if len(spec.ReturnValue) == 0 {
+		return errors.New("value is required")
+	}
+
+	return nil
+}
+
+func validateJVMStressAction(spec *JVMStressSpec) error {
+	if spec.CPUCount == 0 && len(spec.MemoryType) == 0 {
+		return errors.New("one of cpu-count and mem-type is required")
+	}
+
+	if spec.CPUCount > 0 && len(spec.MemoryType) > 0 {
+		return errors.New("inject stress on both CPU and memory is not support")
+	}
+
+	return nil
+}
+
+func validateJVMRuleDataAction(spec *JVMRuleDataSpec) error {
+	if len(spec.RuleData) == 0 {
+		return errors.New("rule-data is required")
+	}
+
+	return nil
+}
+
+func validateClockAction(spec *ClockSpec) error {
+	if spec.Pid == 0 {
```

Comment:
How about checking the range of pid?

Reply:
done, only check whether the PID is less than 0, because users can update the max PID in Linux